tokio-postgres/src/simple_query.rs

```diff
@@ -85,35 +85,36 @@ impl Stream for SimpleQueryStream {
 
     fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         let this = self.project();
-        loop {
-            match ready!(this.responses.poll_next(cx)?) {
-                Message::CommandComplete(body) => {
-                    let rows = extract_row_affected(&body)?;
-                    return Poll::Ready(Some(Ok(SimpleQueryMessage::CommandComplete(rows))));
-                }
-                Message::EmptyQueryResponse => {
-                    return Poll::Ready(Some(Ok(SimpleQueryMessage::CommandComplete(0))));
-                }
-                Message::RowDescription(body) => {
-                    let columns = body
-                        .fields()
-                        .map(|f| Ok(SimpleColumn::new(f.name().to_string())))
-                        .collect::<Vec<_>>()
-                        .map_err(Error::parse)?
-                        .into();
+        match ready!(this.responses.poll_next(cx)?) {
+            Message::CommandComplete(body) => {
+                let rows = extract_row_affected(&body)?;
+                Poll::Ready(Some(Ok(SimpleQueryMessage::CommandComplete(rows))))
+            }
+            Message::EmptyQueryResponse => {
+                Poll::Ready(Some(Ok(SimpleQueryMessage::CommandComplete(0))))
+            }
+            Message::RowDescription(body) => {
+                let columns: Arc<[SimpleColumn]> = body
+                    .fields()
+                    .map(|f| Ok(SimpleColumn::new(f.name().to_string())))
+                    .collect::<Vec<_>>()
+                    .map_err(Error::parse)?
+                    .into();
 
-                    *this.columns = Some(columns);
-                }
-                Message::DataRow(body) => {
-                    let row = match &this.columns {
-                        Some(columns) => SimpleQueryRow::new(columns.clone(), body)?,
-                        None => return Poll::Ready(Some(Err(Error::unexpected_message()))),
-                    };
-                    return Poll::Ready(Some(Ok(SimpleQueryMessage::Row(row))));
-                }
-                Message::ReadyForQuery(_) => return Poll::Ready(None),
-                _ => return Poll::Ready(Some(Err(Error::unexpected_message()))),
+                *this.columns = Some(columns.clone());
+                Poll::Ready(Some(Ok(SimpleQueryMessage::RowDescription(
+                    columns.clone(),
```

Comment:
You shouldn't need to clone here.

Reply:
I'm confused on how to best avoid this clone. Previously, it was able to move the Arc<[SimpleColumn]> over to *this.columns because the Arc wasn't shared, but now since it has to go into the RowDescription and *this.columns, it can no longer move the Arc (from my understanding).

The only way thus far I've been able to avoid 2 clones has been to do:
```
*this.columns = Some(columns);
Poll::Ready(Some(Ok(SimpleQueryMessage::RowDescription(
    this.columns.as_mut().unwrap().clone(),
))))
```

Which is more opaque to me, but I might be missing some benefit to this or that might be the preferred style.

Reply:
I was originally confused about this, but thought it over and figured out a solution. I've added a change to reduce it to just 1 clone instead of 2.